proposals/004-routing-api.md

```diff
@@ -0,0 +1,248 @@
+
+```
+                           my-backing-cluster
+                          ^
+                         /
+                        / foo
+     requests          /
+    ---------> my-router
+                       \
+                        \ bar              
+                         \                ...
+                          v              /
+                           my-other-router --- ...
+                                         \
+                                          ...
+```
+
+All _possible_ routes through the graph can be determined statically from the proxy configuration, but the routing of any individual incoming request is determined at runtime
+can might involve multiple outgoing requests from any given router.
+Validation performed at proxy startup will reject cyclic graphs.
+This will prevent the possibility of a request getting stuck in a router loop. 
+
+In order for non-trivial router graphs to be useful, `Router` authors will need to follow the same principle of composition as `Filter` authors.
+That is, a `Router` implementation should only talk to its receivers, and not, for example, make direct connections of its own to a backing cluster.
+To do so would prevent use of that router implementation in a larger graph. 
+
+The `cluster` propety names a network-reachable backing cluster that speaks the Kafka procotol. It has the same schema as the `targetCluster` property of a virtual cluster.
+
+The existing virtual cluster schema will be modified to support top level `clusters` and to make use of `routers`.
+
+Specifically:
+
+* the existing `targetCluster` property will be made optional, and deprecated
+* a new `cluster` property will support referencing a target cluster by name (using a distict property name seems slightly nicer than overloading the allowed type of the existing `targetCluster` to support `string` or `object`).
+* support for new `router` property will be added. This is a reference to a router defined in `routerDefinitions`. 
+* exactly one of `router`, `cluster` or `targetCluster` will be required.
+* `router` is mutually exclusive with `filters`. 
+
+For example the old-style:
+
+```yaml
+virtualClusters:
+  - name: my-vc
+    portIdentifiesNode: ...
+    filters: 
+      ...
+    targetCluster: 
+      bootstrapServer: ...
+      tls: ...
+``` 
+
+would be rewritten:
+
+```yaml
+clusters:
+  - name: my-backing-cluster
+    bootstrapServer: ...
+    tls: ...
+virtualClusters:
+  - name: my-vc
+    portIdentifiesNode: ...
+    filters: 
+      ...
+    cluster: my-backing-cluster
+``` 
+
+An example of the `router` functionality:
+
+```yaml
+clusters:
+  - name: my-backing-cluster
+    bootstrapServer: ...
+    tls: ...
+routerDefinitions:
+  - name: my-router
+    type: MyRouter
+    config: 
+      ...
+    routes:
+    - name: to-backing-cluster
+      filters: # a list of filter names
+        ...
+      cluster: my-backing-cluster
+virtualClusters:
```

Comment:
Strictly speaking, is the notion of virtual clusters even needed at this point? IIUC, it's equivalent to having a router definition with a no-op router and one single defined route with the filter chain. There probably is value in having this more concise way of modeling that common case, but it also goes a bit against the principle of providing exactly one way to accomplish a given task. Some food for thought, not sure what's best.